apps/src/codegen.js

```diff
@@ -677,6 +671,20 @@ exports.selectCurrentCode = function (interpreter,
   var userCodeRow = -1;
   if (interpreter && interpreter.stateStack[0]) {
     var node = interpreter.stateStack[0].node;
+
+    if (node.type === 'ForStatement') {
+      var mode = interpreter.stateStack[0].mode || 0;
+      if (mode === 0) {
+        node = node.init;
+      } else if (mode === 1) {
+        node = node.test;
+      } else if (mode === 2) {
+        node = node.body;
+      } else if (mode === 3) {
+        node = node.update;
+      }
+    }
+
```

Comment:
This seems pretty safe. It allows us to be more fine grained in selecting code. We could conceivably do this for other node types that have multiple states, (function calls, etc.).

It does mean (for the purposes of breakpoints and anything else that counts on `userCodeRow`) that moving into different parts of the `ForStatement` could now return different row values (since we previously always used the `node.start` to determine the row, and now we're using the sub-node's `start` value).